Signup: convert getUsernameSuggestion to action thunk, don't read Redux store from context

jsnajdr opened the pull request:
Reading `store` from React context is an undocumented API that breaks with React Redux 7.
Let's convert the `getUsernameSuggestion` function to a Redux action thunk. That's a much
more mainstream way to get access to Redux store methods.

This is a spinoff PR from #32129, the big PR that upgrades React Redux to v7. This change can be merged independently.

**How to test:**
Go through a flow that has `domains` step first and the `user` at the end. The only current one is `desktop`.

Verify that after picking a domain, a suggested username is stored in `state.signup.optionalDependencies.suggestedUsername` and that it's pre-filled in the `user` step form.

jsnajdr committed:
Signup: convert getUsernameSuggestion to action thunk, don't read Redux store from context

Reading `store` from React context is an undocumented API that breaks with React Redux 7.
Let's convert the `getUsernameSuggestion` function to a Redux action thunk. That's a much
more mainstream way to get access to Redux store methods.

p-jackson commented on client/signup/steps/domains/actions.js:
```diff
@@ -0,0 +1,67 @@
+		data: resultingUsername,
+	} );
+
+	return resultingUsername;
```
The previous callback version of the action didn't have this return value. And it's never used this way, the suggested username is always read from the store.

Maybe the action creator would be better called `fetchUsernameSuggestion`?

p-jackson reviewed (commented):
Excited to be getting rid of these gnarly uses of the `store` straight out of context!

I appreciate splitting up `step-actions.js` some, it's getting pretty big! I don't think I'd file this action creator under `steps/domains/actions.js` though as the action and state isn't very domain specific.

What do you think of `/client/state/signup/optional-dependencies/actions.js` which would keep the reducers/selectors/etc. grouped together?

jsnajdr commented:
@p-jackson Thanks for the review comments, I implemented all of them:
- renamed the action to `fetchUsernameSuggestion`
- moved to `state/signup/optional-dependencies/action`

> The previous callback version of the action didn't have this return value.

That's because functions with callbacks don't return anything. But promise-returning functions can return something useful.

In case of async Redux actions, it can be convenient to receive the result as a promise:
```js
const suggestion1 = await dispatch( fetchUsernameSuggestion() );
const suggestion2 = getSuggestedUsername( getState() );
```
Both variables have exactly the same value. `suggestion1` used a convenient shortcut to get it. Very useful in code that combines Redux actions and local state into a hybrid async flow.

We don't use the return value anywhere at this moment, but it's a nice convention to follow.

p-jackson reviewed (approved):
Thanks @jsnajdr, tests well, LGTM
